File: labs/05-rate-limiter/python/main.py
import threading
import time

class RateLimiter:
    def __init__(self, capacity, refill_amount, refill_interval, clock=time.monotonic):
        if capacity <= 0:
            raise ValueError("Capacity must be greater than 0")
        if refill_amount <= 0:
            raise ValueError("Refill amount must be greater than 0")
        if refill_interval <= 0:
            raise ValueError("Refill interval must be greater than 0")
        self.capacity = capacity
        self.tokens = capacity
        self.refill_amount = refill_amount
        self.refill_interval = refill_interval
        self.clock = clock
        self.last_refill = self.clock()
        self.lock = threading.Lock()

    def allow(self):
        with self.lock:
            current_time = self.clock()
            elapsed_time = current_time - self.last_refill
            if elapsed_time >= self.refill_interval:
                refill_count = int(elapsed_time / self.refill_interval)
                self.tokens = min(self.capacity, self.tokens + refill_count * self.refill_amount)
                self.last_refill += refill_count * self.refill_interval

            if self.tokens > 0:
                self.tokens -= 1
                return True
            else:
                return False
        

File: labs/05-rate-limiter/go/main.go
package main

import (
	"fmt"
	"sync"
	"time"
)

type Clock func() time.Time

type RateLimiter struct {
	mu             sync.Mutex
	capacity       int
	tokens         int
	refillAmount   int
	refillInterval time.Duration
	clock          Clock
	lastRefillTime time.Time
}

func NewRateLimiter(capacity int, refillAmount int, refillInterval time.Duration, clock Clock) (*RateLimiter, error) {
	if capacity <= 0 {
		return nil, fmt.Errorf("capacity must be a positive integer")
	}
	if refillAmount <= 0 {
		return nil, fmt.Errorf("refillAmount must be a positive integer")
	}
	if refillInterval <= 0 {
		return nil, fmt.Errorf("refillInterval must be a positive duration")
	}
	if clock == nil {
		return nil, fmt.Errorf("clock must not be nil")
	}

	return &RateLimiter{
		capacity:       capacity,
		tokens:         capacity,
		refillAmount:   refillAmount,
		refillInterval: refillInterval,
		clock:          clock,
		lastRefillTime: clock(),
	}, nil
}

func (rl *RateLimiter) Allow() bool {
	rl.mu.Lock()
	defer rl.mu.Unlock()

	now := rl.clock()
	elapsed := now.Sub(rl.lastRefillTime)

	if elapsed >= rl.refillInterval {
		refills := int(elapsed / rl.refillInterval)
		rl.tokens += refills * rl.refillAmount
		if rl.tokens > rl.capacity {
			rl.tokens = rl.capacity
		}
		rl.lastRefillTime = rl.lastRefillTime.Add(time.Duration(refills) * rl.refillInterval)
	}

	if rl.tokens > 0 {
		rl.tokens--
		return true
	}
	return false
}
